oauth2/oauth2_refresh_token_test.go

```diff
@@ -0,0 +1,158 @@
+	_, err = postgresRegistry.OAuth2Storage().GetRefreshTokenSession(ctx, tokenSignature, nil)
+	require.NoError(t, err)
+	provider := postgresRegistry.OAuth2Provider()
+
+	var wg sync.WaitGroup
+	for run := 0; run < testRuns; run++ {
+		barrier := make(chan struct{})
+		errorsCh := make(chan error, workers)
+
+		go func() {
+			for w := 0; w < workers; w++ {
+				wg.Add(1)
+				go func(run, worker int) {
+					defer wg.Done()
+					ctx, _ := context.WithDeadline(context.Background(), time.Now().Add(5*time.Second))
+					// all workers will block here until the for loop above has launched all the worker go-routines
+					// this is to ensure we fire all the workers off at the same
+					<-barrier
+					_, err := provider.NewAccessResponse(ctx, request)
+					errorsCh <- err
+				}(run, w)
+			}
+
+			// wait until all workers have completed their work
+			wg.Wait()
+			close(errorsCh)
+		}()
+
+		// let the race begin!
+		// all worker go-routines will now attempt to hit the "NewAccessResponse" method
+		close(barrier)
+
+		// process worker results
+
+		// successCount is the number of workers that were able to call "NewAccessResponse" without receiving an error.
+		// if the successCount at the end of a test run is bigger than one, it means that multiple access/refresh tokens
+		// were issued using the same refresh token! - https://knowyourmeme.com/memes/scared-hamster
+		var successCount int
+		for err := range errorsCh {
+			if err != nil {
+				switch cause := errorsx.Cause(err).(type) {
+				case *fosite.RFC6749Error:
+					switch cause.Name {
+					case fosite.ErrInvalidRequest.Name:
+						assert.Equal(t, fosite.ErrInvalidRequest.Description, cause.Description)
+						assert.Contains(t, cause.Debug, "could not serialize access due to concurrent update (SQLSTATE 40001)")
+					default:
+						t.Errorf("an unexpected RFC6749 error with the name %q was returned.\n"+
+							"Hint: has the refresh token error handling changed in fosite? If so, you need to add further "+
+							"assertions here to cover the additional errors that are being returned by the handler.\n\n"+
+							"Error description: %s\n"+
+							"Error debug: %s\n"+
+							"Error hint: %s\n"+
+							"Raw error: %+v",
+							cause.Name,
+							cause.Description,
+							cause.Debug,
+							cause.Hint,
+							err)
+					}
+				default:
+					t.Errorf("expected underlying error to be of type '*fosite.RFC6749Error', but it was "+
+						"actually of type %T: %+v", err, err)
+				}
+			} else {
+				successCount++
+			}
+		}
+
+		assert.Equal(t, 1, successCount, "CRITICAL: in test iteration %d, %d out of %d workers "+
+			"were able to use the refresh token. Exactly ONE was expected to be have been successful.",
+			run,
+			successCount,
+			workers)
+
+		// reset state for the next test iteration
+		_ = postgresRegistry.OAuth2Storage().RevokeRefreshToken(ctx, request.ID)
+		_ = postgresRegistry.OAuth2Storage().CreateRefreshTokenSession(ctx, tokenSignature, request)
+	}
+}
```

Comment:
@aeneasr if you have a spare cycle, it might be good to get you to try running this test locally so you get a feel for the value it provides. 

🎩 instructions:

- `git fetch`
- `git checkout aaslamin:fix/1735-and-1719`
- `cd <to hydra dir>`
- `cd oauth2`
- start docker
- `go test -race -run=TestCreateRefreshTokenSessionStress -count=1 -v`

Now, just to see what happens when something breaks:

Go to https://github.com/ory/hydra/pull/1766/files#diff-3d061c6808115be018958df459c564e2R465 and change the consistency level to `sql.LevelReadCommitted` and re-run the test. Can we make the failure experience better for the developer debugging the test?